### PR TITLE
2020-03-26 MaterialIcons proposed implementation

### LINKED ISSUE #53: Support additional icon option, notably Material & Open Iconic
**Is your feature request related to a problem? Please describe.**
I have two projects in which I would like to use Blazored.Toast. They make use of the Open Iconic Glyphs and Material Design icons. My preference is not to add FontAwesome just to get icons on toast popups.

**Describe the solution you'd like**
I believe that adding parameters to BlazoredToasts.razor.cs would be an effective way of specifying icons from these two sources. I've implemented a POC by adding InfoIconMatName, ErrorIconMatName (plus success and warning). I extended the ToastSettings to include IconMatName as a property and updated the constructor. Finally, in BlazorToast.razor I added
```
        @if (!string.IsNullOrWhiteSpace(ToastSettings.IconMatName))
        {
            <div class="blazored-toast-icon">
                <i class="material-icons mdc-button__icon" aria-hidden="true">@ToastSettings.IconMatName</i>
            </div>
        }
```
below the existing icon code.

**Describe alternatives you've considered**
An alternative would be to add an IconType for each of the four states as an enumeration of Class, MatIcon, OpenIconic and make the use of IconClass be the class if type class, or the font selector if MatIcon or OpenIconic. It feels a little funny to use the className for the icon name but it beats having a breaking change to get a better name IMO.

### PR DESCRIPTION
**Pull request by MarkStega**
Resolves #53 

Very modest changes in 6 files. I demonstrate MD icons in the BlazorServerSideSample and left BlazorClientSideSample using the FA icons. Using GitKraken the dif on BlazoredToast.razor looks like I changed lots where, in fact, it was formatted in VS changing indentation and the only real change was the addition of the IconName.

I have not yet changed the README.md pending discussion of the proposed changes...

### CODE REVIEW

**Comment by chrissainty** [comment]
Hi Mark,

I've had a quick look at your implementation and it should work fine. I'm just wondering if we could be a bit slicker about it.

I wonder if we introduce an `IconType` parameter to the component which is an enum with values for the various icon packs we support. Based on which enum value is picked we can apply the icon class in the correct place in the markup for that icon pack and save having to have `WarningIconClass` and `WarningIconName`.

Does that make sense?

So in terms of API I would do something like this:

**Material Icons**
```razor
<BlazoredToasts Position="ToastPosition.BottomRight"
                           Timeout="10"
                           IconType="IconType.Material"
                           SuccessClass="success-toast-override"
                           SuccessIcon="done_outline" />
```

**Fontawesome Icons**
```razor
<BlazoredToasts Position="ToastPosition.BottomRight"
                           Timeout="10"
                           IconType="IconType.Fontawesome"
                           SuccessClass="success-toast-override"
                           SuccessIcon="fa fa-thumbs-up" />
```

**Comment by MarkStega** [comment]
I can easily do that -- Good suggestion

**Comment by MarkStega** [comment]
Chris, Per your suggestion the api has been changed. I wonder if the IconType enumeration should be {Class, Material} rather then (FontAwesome, Material) to better indicate that any icon set the specifies the icon by class name can be used?

The client example shows FA icons, the server example shows MD fonts.

I took a shot at the Readme.md update.

File: src/Blazored.Toast/Blazored.Toast.csproj [hunk 1]
```diff
@@ -16,7 +16,7 @@
     <PackageTags>Blazored Blazor Razor Components Toast Popup Notification</PackageTags>
     <Description>A JavaScript free Toast library for Blazor and Razor Components applications.</Description>
     <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
-    <Version>2.0.9</Version>
+    <Version>2.0.10-Preview2</Version>
```
Comment (chrissainty) [inline src/Blazored.Toast/Blazored.Toast.csproj:19]: Could you back this change out, we'll sort version numbers when we release.

File: src/Blazored.Toast/BlazoredToasts.razor.cs [hunk 2]
```diff
@@ -8,18 +8,21 @@
 
 namespace Blazored.Toast
 {
+    public enum IconType { FontAwesome, Material };
+
     public partial class BlazoredToasts
     {
         [Inject] private IToastService ToastService { get; set; }
 
+        [Parameter] public IconType IconType { get; set; } = IconType.FontAwesome;
```
Comment (chrissainty) [inline src/Blazored.Toast/BlazoredToasts.razor.cs:17]: Should we default this? I'm not sure we should assume FontAwesome is the default icon choice. Maybe we should have it as a required parameter with a null check.

File: src/Blazored.Toast/Configuration/ToastSettings.cs [hunk 3]
```diff
@@ -2,19 +2,27 @@
...
+            string additionalClasses,
+            string iconName)
         {
             Heading = heading;
             Message = message;
+            IconType = iconType;
             BaseClass = baseClass;
             AdditionalClasses = additionalClasses;
-            IconClass = iconClass;
+            IconSelector = iconName;
```
Comment (chrissainty) [inline src/Blazored.Toast/Configuration/ToastSettings.cs:18]: I think could just be _icon_, I think it's clear enough and we're not dancing around if its a class for an icon or a string identifier.

```suggestion
            Icon = icon;
```

File: src/Blazored.Toast/Configuration/ToastSettings.cs [hunk 4]
```diff
@@ -2,19 +2,27 @@
...
         }
 
         public string Heading { get; set; }
         public string Message { get; set; }
         public string BaseClass { get; set; }
         public string AdditionalClasses { get; set; }
-        public string IconClass { get; set; }
+        public string IconSelector { get; set; }
```
Comment (chrissainty) [inline src/Blazored.Toast/Configuration/ToastSettings.cs:25]:
```suggestion
        public string Icon { get; set; }
```

File: src/Blazored.Toast/Blazored.Toast.csproj (hunk 1 above)
Comment (MarkStega) [reply src/Blazored.Toast/Blazored.Toast.csproj:19]: Done - Sorry that slipped in -- I publish to a local directory to use the project and I need the larger version number to drive an update.

File: src/Blazored.Toast/BlazoredToasts.razor.cs (hunk 2 above)
Comment (MarkStega) [reply src/Blazored.Toast/BlazoredToasts.razor.cs:17]: Before I changed the names of the parameters this kept everything working as it did originally (It was done with a different parameter but the idea was the same). Your call, I'm pretty agnostic on this.

File: src/Blazored.Toast/Configuration/ToastSettings.cs (hunk 4 above)
Comment (MarkStega) [reply src/Blazored.Toast/Configuration/ToastSettings.cs:25]: Done

**Comment by MarkStega** [comment]
I'll do another push if you want to not default the icon type...

File: src/Blazored.Toast/Configuration/ToastSettings.cs (hunk 3 above)
Comment (MarkStega) [reply src/Blazored.Toast/Configuration/ToastSettings.cs]: Done

File: src/Blazored.Toast/BlazoredToasts.razor.cs (hunk 2 above)
Comment (chrissainty) [reply src/Blazored.Toast/BlazoredToasts.razor.cs:17]: I think now we've changed the API we might as well go all in. By making the `IconType` mandatory it makes things consistent.

Comment (MarkStega) [reply src/Blazored.Toast/BlazoredToasts.razor.cs]: Done; I had to add the errorUI html to the html startup pages in both examples. In SSB it displays the error as expected. In CSB the error briefly flashes and then the app continues and gets rendered. The null IconType prevents the icons from displaying. I observed the continuation without the error message looking at the console in CSB (that's what made me add the errorUI). I think that this is a CSB issue with wasm. I will raise an issue on the Asp.Net repository if you don't have any special insight.

Comment (MarkStega) [reply src/Blazored.Toast/BlazoredToasts.razor.cs]: For you to look at this just take the IconType parameter out of the client side example.

**Comment by MarkStega** [comment]
Proving that I can still get confused by github, I resolved a conflict where github didn't like my change from a single long line to parameters spaced line by line. When I committed that change a full merge of master into the PR branch occurred. I inspected each merged file and it appears to be correct (preserving the changes making the message a RenderFragment rather than a string).

**Comment by chrissainty** [comment]
We got there @MarkStega! 😃

**Pull request merged**